build/controller_gen.bzl

```diff
@@ -0,0 +1,78 @@
+# Copyright 2019 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# TODO: Move this to Kubebuilder repository
+
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
+load("@io_kubernetes_build//defs:go.bzl", "go_genrule")
+load("//build:install.bzl", "install")
+
+CONTROLLER_GEN = "//vendor/sigs.k8s.io/controller-tools/cmd/controller-gen"
+
+def _qualified_genfile(label):
+  return "$$GO_GENRULE_EXECROOT/$(location %s)" % label
+
+def controller_gen(name, importpath, api, visibility, deps = []):
```

Comment:
I would love to bring these upstream if that's even possible. We should also add some more comments around these functions and their objective.

Reply:
100% agree. Have added comments for now, then I'll PR to the controller-runtime repo after we know we're happy with it.